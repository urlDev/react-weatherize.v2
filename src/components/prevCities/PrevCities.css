.localStorage {
    display: block;
    /* width: 10%; */
    background: none;
    text-align: left;
    cursor: pointer;
    font-family: 'Roboto', sans-serif;
    color: white;
    outline: none;
    margin: 0;
}

.localContainer {
    width: 80%;
    height: 20%;
    padding: 0;
    margin: 60px auto 30px auto;
}

.lengthTwo button {
    margin-bottom: 20px;
}

.lengthMore {
    display: flex;
    justify-content: space-between;
    flex-direction: column;
}

@media only screen and (min-width: 900px) {
    .localContainer {
        margin-top: 10px;
    }
}

@media (min-width: 800px) and (min-height: 800px) {
    .localContainer {
        margin-top: -20vh;
    }
}