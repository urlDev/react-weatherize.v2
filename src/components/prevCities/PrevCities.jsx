import React from 'react';
import {
  useSelector,
  useDispatch,
} from '../../tests/__mocks__/reactReduxHooks';

import { fetchWeather } from '../../redux/weather/weatherActions';

import './PrevCities.css';

/*
Because of a bug that got created after mockStore/testing,
I had to add storage.storage and || operator.
*/

const PrevCities = () => {
  const storage = useSelector((state) => state.weather.storage);
  const dispatch = useDispatch();

  return (
    <div
      className={
        (storage.storage || storage).length === 2
          ? 'localContainer lengthTwo'
          : 'localContainer lengthMore'
      }
    >
      {(storage.storage || storage).slice(-3).map((city) => (
        <button
          type="button"
          className="localStorage"
          onClick={() => {
            dispatch(fetchWeather(city));
          }}
          key={city}
        >
          {city}
        </button>
      ))}
    </div>
  );
};

export default PrevCities;
