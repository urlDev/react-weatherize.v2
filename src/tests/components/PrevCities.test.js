/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { shallow } from 'enzyme';

import PrevCities from '../../components/prevCities/PrevCities';

import * as ReactReduxHooks from '../__mocks__/reactReduxHooks';
import { mockStore } from '../__mocks__/store';
import defaultCity, { lengthTwo } from '../fixtures/storage';

describe('Previous Cities component', () => {
  let wrapper;
  let store;

  test('Should render previous cities on mount correctly', () => {
    store = mockStore({
      loading: false,
      error: null,
      weather: [],
      storage: defaultCity,
    });
    jest
      .spyOn(ReactReduxHooks, 'useSelector')
      .mockImplementation((state) => store.getState());

    jest
      .spyOn(ReactReduxHooks, 'useDispatch')
      .mockImplementation(() => store.dispatch);

    wrapper = shallow(<PrevCities store={store} />);
    expect(wrapper).toMatchSnapshot();
    expect(store.getState().storage.length).toBe(1);
  });

  test('Should render previous cities correctly after searching for more than one city', () => {
    store = mockStore({
      loading: false,
      error: null,
      weather: [],
      storage: lengthTwo,
    });
    jest
      .spyOn(ReactReduxHooks, 'useSelector')
      .mockImplementation((state) => store.getState());

    jest
      .spyOn(ReactReduxHooks, 'useDispatch')
      .mockImplementation(() => store.dispatch);

    wrapper = shallow(<PrevCities store={store} />);

    expect(wrapper).toMatchSnapshot();
    expect(store.getState().storage.length).toBe(2);
  });
  // There is a weird bug connecting mockStore with store(or persisted store).
  // After running tests, its best to delete local storage and refresh the page
  // Tests added another layer of state (e.g. storage.storage, weather.weather)
  // Trying to find the reason
});
