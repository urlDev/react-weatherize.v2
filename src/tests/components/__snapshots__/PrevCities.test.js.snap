// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Previous Cities component Should render previous cities correctly after searching for more than one city 1`] = `
<div
  className="localContainer lengthTwo"
>
  <button
    className="localStorage"
    key="Helsinki"
    onClick={[Function]}
    type="button"
  >
    Helsinki
  </button>
  <button
    className="localStorage"
    key="Istanbul"
    onClick={[Function]}
    type="button"
  >
    Istanbul
  </button>
</div>
`;

exports[`Previous Cities component Should render previous cities on mount correctly 1`] = `
<div
  className="localContainer lengthMore"
>
  <button
    className="localStorage"
    key="Helsinki"
    onClick={[Function]}
    type="button"
  >
    Helsinki
  </button>
</div>
`;
